### ISSUE TITLE
custom_stopping_strings not working as intended

### ISSUE THREAD

**Issue by FJ05** [oobabooga/text-generation-webui#1369]
### Describe the bug

custom_stopping_strings not working as intended when used in api mode.

If I understand the function of custom_stopping_strings right it's supposed to halt the generation when encountering a string in the array.

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Reproduction

```javascript
var custom_stopping_strings = ["\n", message.author.username + " says:", "BotTerfly says:"];
``` 
```javascript
return new Promise(function (resolve) {
		const data = JSON.stringify({
			prompt : inputPrompt,
			do_sample : false,
			temperature : 0.99,
			top_p : 0.9,
			typical_p : 1,
			repetition_penalty : 1.1,
			encoder_repetition_penalty : 1.0,
			top_k : 40,
			min_length : 0,
			no_repeat_ngram_size : 0,
			num_beams : 1,
			penalty_alpha : 0,
			length_penalty : 1,
			early_stopping : true,
			seed : -1,
			add_bos_token : true,
			custom_stopping_strings : custom_stopping_strings,
			truncation_length : 2048,
			ban_eos_token : false,
			skip_special_tokens : true,
			stop_at_newline : true
		});
		  
		const options = {
			hostname: oobServer,
			port: oobPort,
			path: '/api/v1/generate', 
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
				'Content-Length': data.length
			}
		};
		console.log(data);
		const req = http.request(options, (res) => {
			let responseData = '';
		
			res.on('data', (chunk) => {
			responseData += chunk;
			});
		
			res.on('end', () => {
				var response = JSON.parse(responseData);
				resolve(response.results[0].text);
			});
		});
		  
		req.on('error', (error) => {
			console.error('Error:', error);
		});
		
		req.write(data);
		req.end();
	});
``` 

### Screenshot

![image](https://user-images.githubusercontent.com/48802078/233017889-18848f8a-ffeb-4652-ae53-b76009ce80bf.png)

### Logs

```shell
I have no logs but the output is in the screenshot
```

### System Info

```shell
Windows 11 22H2 (os Build 22624.1610)
Nvidia RTX 2070 super
AMD ryzen 2700x
```

**Comment by DeSinc** [comment]
Yep I'm seeing same issues. I had a look at the api.py changes and reverted that code but that didn't seem to solve it, at least not the very most recent changes made 3 days ago that included the defaults in the param list.

**Comment by oobabooga** [comment]
I think that this should have fixed it: https://github.com/oobabooga/text-generation-webui/commit/9d9ae6293833ce31bbb5ed5d9a04b033d1e3896d

Note that sometimes you have to add a leading space to the stopping string for it ti work. For instance, `" my name is"` instead of `"my name is"`

**Comment by FJ05** [comment]
I updated all the files and added the leading spaces for each string in the array as you told me
```javascript
var custom_stopping_strings = ["\n"," " + message.author.username + " says:", " BotTerfly says:"];
``` 
But when sending data to the api using /api/v1/generate custom_stopping_strings still won't work.

Is it just me being stupid or is there still something wrong with the api?

**Comment by oobabooga** [comment]
Use the api example as a reference, you might be sending the wrong parameter names

https://github.com/oobabooga/text-generation-webui/blob/main/api-example.py#L42

**Comment by FJ05** [comment]
I used the reference making sure I don't have any wrong parameter names. All other parameters seems to work just fine except stopping_strings.

Here is a snippet of the code:
```javascript
const data = JSON.stringify({
	'prompt': inputPrompt,
	'max_new_tokens': 200,
	'do_sample': true,
	'temperature': 0.72,
	'top_p': 0.73,
	'typical_p': 1,
	'repetition_penalty': 1.1,
	'encoder_repetition_penalty': 1.0,
	'top_k': 0,
	'min_length': 0,
	'no_repeat_ngram_size': 0,
	'num_beams': 1,
	'penalty_alpha': 0,
	'length_penalty': 1,
	'early_stopping': false,
	'seed': -1,
	'add_bos_token': true,
	'truncation_length': 2048,
	'ban_eos_token': false,
	'skip_special_tokens': true,
	'stopping_strings': custom_stopping_strings,
});
		
const options = {
	hostname: oobServer,
	port: oobPort,
	path: '/api/v1/generate',
	method: 'POST',
	headers: {
		'Content-Type': 'application/json',
		'Content-Length': data.length
	}
};
const req = http.request(options, (res) => {
	let responseData = '';

	res.on('data', (chunk) => {
	responseData += chunk;
	});

	res.on('end', () => {
		var response = JSON.parse(responseData);
		resolve(response.results[0].text);
	});
});
		
req.on('error', (error) => {
	console.error('Error:', error);
	resolve("Error")
});

req.write(data);
req.end();
```

**Comment by DeSinc** [comment]
> I used the reference making sure I don't have any wrong parameter names. All other parameters seems to work just fine except stopping_strings.
> 
> Here is a snippet of the code:
> 
> ```js
> const data = JSON.stringify({
> 	'prompt': inputPrompt,
> 	'max_new_tokens': 200,
> 	'do_sample': true,
> 	'temperature': 0.72,
> 	'top_p': 0.73,
> 	'typical_p': 1,
> 	'repetition_penalty': 1.1,
> 	'encoder_repetition_penalty': 1.0,
> 	'top_k': 0,
> 	'min_length': 0,
> 	'no_repeat_ngram_size': 0,
> 	'num_beams': 1,
> 	'penalty_alpha': 0,
> 	'length_penalty': 1,
> 	'early_stopping': false,
> 	'seed': -1,
> 	'add_bos_token': true,
> 	'truncation_length': 2048,
> 	'ban_eos_token': false,
> 	'skip_special_tokens': true,
> 	'stopping_strings': custom_stopping_strings,
> });
> 		
> const options = {
> 	hostname: oobServer,
> 	port: oobPort,
> 	path: '/api/v1/generate',
> 	method: 'POST',
> 	headers: {
> 		'Content-Type': 'application/json',
> 		'Content-Length': data.length
> 	}
> };
> const req = http.request(options, (res) => {
> 	let responseData = '';
> 
> 	res.on('data', (chunk) => {
> 	responseData += chunk;
> 	});
> 
> 	res.on('end', () => {
> 		var response = JSON.parse(responseData);
> 		resolve(response.results[0].text);
> 	});
> });
> 		
> req.on('error', (error) => {
> 	console.error('Error:', error);
> 	resolve("Error")
> });
> 
> req.write(data);
> req.end();
> ```

Guys, you're talking past each other. He's using the extensions API like me. The extensions API takes the parameter custom_stopping_ strings. It does not listen to stopping_strings. This is not a parameter. It doesn't exist. I already found this out the hard way after trying this non-working parameter for days before I randomly found out from some random person on the internet that it's actually called custom_stopping_strings.

The problem is even though this was working a few days ago (extensions API like he's using, NOT the default API), a recent update within the last few days _has_ broken it. It was working and now it is not, and I can verify it with 100% certainty. The extensions API parameter custom_stopping_strings was working and now it is not.

**Comment by oobabooga** [comment]
Can you see if the API extension works after this change? https://github.com/oobabooga/text-generation-webui/commit/96ba55501ca3cd22ccaa5782facea5f11b4669cf

Name the parameter `stopping_strings` this time.

**Comment by int19h** [comment]
I was just looking at this recently, and so far as I can tell, the implementation of stopping strings via `_SentinelTokenStoppingCriteria` makes "stopping strings" misnomer, because the implementation actually matches *tokens*, not strings. So your stopping string gets converted to the sentinel token list first, and then that gets matched against the output. This works consistently if the mapping is unique both ways, but in many cases the same character can be represented in the output by different tokens.

For example, if you run LLaMA and set it to stop at "a", that doesn't help you if the model outputs "aa", because the latter is a distinct token. Or, more realistically, if you ask it to stop at e.g. "}" (for generating JSON), it won't stop on " }" or "}}" or "},". Or conversely if you set it to "}}" the model might actually output it as two "}" tokens.

**Comment by oobabooga** [comment]
That is correct, @int19h.

**Comment by deathward4** [comment]
I've pulled the latest version, and I'm still having this issue. I'm using the extended API like the others in the thread.
The behavior is highly unusual because my custom stop strings work when entered into the gradio UI, BUT when calling from the extended API (using either or in some attempts, both stopping_strings and custom_stopping_strings) it does not work and continues to run until it runs out of tokens. If this was an issue with how the actual "stopping strings" are defined, it seems unusual for the same strings to work in the WebUI but not the API.

**Comment by raziel2001au** [comment]
I'm not sure if this is related, but I pulled around an hour ago and now I'm getting this stack trace when connecting to oobabooga from TavernAI (was working last night so the update seems like the likely candidate(:
```
Exception occurred during processing of request from ('127.0.0.1', 64422)
Traceback (most recent call last):
  File "D:\Development\text-generation-webui\installer_files\env\lib\socketserver.py", line 683, in process_request_thread
    self.finish_request(request, client_address)
  File "D:\Development\text-generation-webui\installer_files\env\lib\socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "D:\Development\text-generation-webui\installer_files\env\lib\socketserver.py", line 747, in __init__
    self.handle()
  File "D:\Development\text-generation-webui\installer_files\env\lib\http\server.py", line 433, in handle
    self.handle_one_request()
  File "D:\Development\text-generation-webui\installer_files\env\lib\http\server.py", line 421, in handle_one_request
    method()
  File "D:\Development\text-generation-webui\text-generation-webui\extensions\api\script.py", line 67, in do_POST
    for a in generator:
  File "D:\Development\text-generation-webui\text-generation-webui\modules\text_generation.py", line 195, in generate_reply
    for st in (stopping_strings, ast.literal_eval(f"[{state['custom_stopping_strings']}]")):
KeyError: 'custom_stopping_strings'
```

**Comment by raziel2001au** [comment]
I wouldn't know, I'm just using Tavern AI... Not hitting it manually. I would expect it to keep working with existing software.

**Comment by FJ05** [comment]
I seem to get the same result as @raziel2001au when running the latest version when trying to use stopping strings. I get the same error message even when the parameter "stopping_strings" is not present in my request. I've also tried replacing the parameter "stopping_strings" with "custom_stopping_strings" but the same error occurs.

```
127.0.0.1 - - [20/Apr/2023 14:11:14] "POST /api/v1/generate HTTP/1.1" 200 -
----------------------------------------
Exception occurred during processing of request from ('127.0.0.1', 61520)
Traceback (most recent call last):
  File "E:\Programming\oobabooga-windows\installer_files\env\lib\socketserver.py", line 683, in process_request_thread
    self.finish_request(request, client_address)
  File "E:\Programming\oobabooga-windows\installer_files\env\lib\socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "E:\Programming\oobabooga-windows\installer_files\env\lib\socketserver.py", line 747, in __init__
    self.handle()
  File "E:\Programming\oobabooga-windows\installer_files\env\lib\http\server.py", line 433, in handle
    self.handle_one_request()
  File "E:\Programming\oobabooga-windows\installer_files\env\lib\http\server.py", line 421, in handle_one_request
    method()
  File "E:\Programming\oobabooga-windows\text-generation-webui\extensions\api\script.py", line 67, in do_POST
    for a in generator:
  File "E:\Programming\oobabooga-windows\text-generation-webui\modules\text_generation.py", line 195, in generate_reply
    for st in (stopping_strings, ast.literal_eval(f"[{state['custom_stopping_strings']}]")):
KeyError: 'custom_stopping_strings'
----------------------------------------
```

**Comment by caretaker0815** [comment]
@permissionBRICK that worked. Many thanks for your workaround

**Comment by oobabooga** [comment]
@raziel2001au are you using Tavern or SillyTavern?

**Comment by raziel2001au** [comment]
Original Tavern AI, latest downloadable exe version for Windows.

**Comment by oobabooga** [comment]
Can you check if this fixed it? https://github.com/oobabooga/text-generation-webui/commit/ff0d0ac552ef69c3bae6bf34397706033d6e2fff

**Comment by raziel2001au** [comment]
I'm unfortunately heading to bed (2:30am), happy to test tomorrow, otherwise if someone who had the same issue can try it out that would be good as well...

**Comment by permissionBRICK** [comment]
> Can you check if this fixed it? [ff0d0ac](https://github.com/oobabooga/text-generation-webui/commit/ff0d0ac552ef69c3bae6bf34397706033d6e2fff)

I just tried it (After reverting the workaround), and it seems to work with Tavern AI.

**Comment by deathward4** [comment]
> Can you check if this fixed it? [ff0d0ac](https://github.com/oobabooga/text-generation-webui/commit/ff0d0ac552ef69c3bae6bf34397706033d6e2fff)

This change fixed it for me! I did have to rename the parameter in my script to use "stopping_strings" instead of custom, but now it works!

**Comment by FJ05** [comment]
stopping_strings works now!
Thanks to everyone for helping and fixing the bug :)
This is the parameters that I am using if anyone is curious:
```javascript
const data = JSON.stringify({
  'prompt': inputPrompt,
  'max_new_tokens': 120,
  'do_sample': true,
  'temperature': 0.72,
  'top_p': 0.73,
  'typical_p': 1,
  'repetition_penalty': 1.1,
  'encoder_repetition_penalty': 1.0,
  'top_k': 0,
  'min_length': 0,
  'no_repeat_ngram_size': 0,
  'num_beams': 1,
  'penalty_alpha': 0,
  'length_penalty': 1,
  'early_stopping': true,
  'seed': -1,
  'add_bos_token': true,
  'truncation_length': 2048,
  'ban_eos_token': false,
  'skip_special_tokens': true,
  'stopping_strings': custom_stopping_strings,
 });
```
